### ISSUE TITLE
Do not sort high cardinality string dictionaries

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#8568]
__Use case:__
This ticket describes making IOx faster for both query and compaction when it has high cardinality individual columns. These columns are common if an id (e.g. `trace_id` or `span_id`)is included as a tag. 

Note this ticket is only referring to cardinality in a *single* column -- it isn't talking about the cardinality of the entire "tag set" (aka the distinct values of all the tags together)

IOx currently stores all tags as the arrow datatype `DataType::Dictionary`

This  works great for low cardinality tags (where there are 10s of distinct values). It does not work well for high cardinality tags where most values are unique. 

__Dictionaries are 40x worse for some High Cardinality Usecases:__

TLDR is that for even moderate sized data (220MB parquet, 5M rows) preserving dictionary encoding for these high cardinality columns is **over 40x** slower  than using `StringArrays`. 

Here is a example that mimics part of what the compactor's work when merging sorted parquet files together:

Unzip [traces.zip](https://drive.google.com/file/d/1SlBGlTrYgcZTITDX_VRG22QZ7IyPX_JR/view?usp=drive_link) 

If we copy the data directly (which preserves the DictionaryEncoding during the plan) it takes **285 seconds (4:45)**:

```shell
datafusion-cli -c "COPY (select attributes, duration_nano, end_time_unix_nano, \"service.name\", \"span.kind\", \"span.name\", span_id, time, trace_id, \"otel.status_code\", parent_span_id from 'traces' ORDER BY time) to 'traces_out.parquet'"

+---------+
| count   |
+---------+
| 5185717 |
+---------+
1 row in set. Query took 285.891 seconds.
```

If we force the high cardinality tags to be `StringArray`s by adding a `::varchar` cast, it takes **between 6.5 and 7.5 seconds.**

```
datafusion-cli -c "COPY (select attributes, duration_nano, end_time_unix_nano, \"service.name\", \"span.kind\", \"span.name\", span_id::varchar, time, trace_id::varchar, \"otel.status_code\", parent_span_id from 'traces' ORDER BY time) to 'traces_out.parquet'"
DataFusion CLI v29.0.0
+---------+
| count   |
+---------+
| 5185717 |
+---------+
1 row in set. Query took 6.585 seconds.
```

__Description of dataset:__

(massive thanks to @JaygeetAtGithub for making this dataset)

The dataset has 5.1M rows in  186 parquet files in a single directory, 231M in total

```
❯ select count(*) from 'traces';
+-----------------+
| COUNT(UInt8(1)) |
+-----------------+
| 5185717         |
+-----------------+
1 row in set. Query took 0.051 seconds.
```

The schema is as follows:
```
describe 'traces';
+--------------------+-----------------------------+-------------+
| column_name        | data_type                   | is_nullable |
+--------------------+-----------------------------+-------------+
| attributes         | Utf8                        | YES         |
| duration_nano      | Int64                       | YES         |
| end_time_unix_nano | Int64                       | YES         |
| service.name       | Dictionary(Int32, Utf8)     | YES         |
| span.kind          | Utf8                        | YES         |
| span.name          | Utf8                        | YES         |
| span_id            | Dictionary(Int32, Utf8)     | YES         |
| time               | Timestamp(Nanosecond, None) | NO          |
| trace_id           | Dictionary(Int32, Utf8)     | YES         |
| otel.status_code   | Utf8                        | YES         |
| parent_span_id     | Utf8                        | YES         |
+--------------------+-----------------------------+-------------+
```

__Proposal:__
TLDR is we propose treating such columns as StringArrays and not DictionaryArrays. I am working on a fuller writeup.

__Alternatives considered:__
Describe other solutions or features you considered.

**Comment by alamb** [comment]
When I profilied the slow case, almost 90% of the time is spent in the `interleave` kernel which is used to construct the output of the sort.

![Screenshot 2023-08-24 at 9 17 00 AM](https://github.com/influxdata/influxdb_iox/assets/490673/ab88b109-3d57-44cb-ba72-4fc000f0b61c)

**Comment by alamb** [comment]
Here is another example query where using dictionary encoding really hurts:

## Input data
use gby_data.zip has two files. `traces.parquet` and `traces_nd.parquet` with the same data, the only difference is that `traces.parquet` is dictionary encoded the two high cardinality tag files:

```shell
$ datafusion-cli -c "describe 'traces.parquet'"
DataFusion CLI v29.0.0
+--------------------+-----------------------------+-------------+
| column_name        | data_type                   | is_nullable |
+--------------------+-----------------------------+-------------+
| attributes         | Utf8                        | YES         |
| duration_nano      | Int64                       | YES         |
| end_time_unix_nano | Int64                       | YES         |
| service.name       | Dictionary(Int32, Utf8)     | YES         |
| span.kind          | Utf8                        | YES         |
| span.name          | Utf8                        | YES         |
| span_id            | Dictionary(Int32, Utf8)     | YES         |
| time               | Timestamp(Nanosecond, None) | NO          |
| trace_id           | Dictionary(Int32, Utf8)     | YES         |
| otel.status_code   | Utf8                        | YES         |
| parent_span_id     | Utf8                        | YES         |
+--------------------+-----------------------------+-------------+
11 rows in set. Query took 0.002 seconds.

$ datafusion-cli -c "describe 'traces_nd.parquet'"
DataFusion CLI v29.0.0
+--------------------+-----------------------------+-------------+
| column_name        | data_type                   | is_nullable |
+--------------------+-----------------------------+-------------+
| attributes         | Utf8                        | YES         |
| duration_nano      | Int64                       | YES         |
| end_time_unix_nano | Int64                       | YES         |
| service.name       | Utf8                        | YES         |
| span.kind          | Utf8                        | YES         |
| span.name          | Utf8                        | YES         |
| span_id            | Utf8                        | YES         |
| time               | Timestamp(Nanosecond, None) | NO          |
| trace_id           | Utf8                        | YES         |
| otel.status_code   | Utf8                        | YES         |
| parent_span_id     | Utf8                        | YES         |
+--------------------+-----------------------------+-------------+
11 rows in set. Query took 0.003 seconds.
```

## Query
Now, running the following query to compute the most recent 10 traces:
```sql
SELECT trace_id, MAX(time) FROM 'traces_nd.parquet' GROUP BY trace_id ORDER BY MAX(time) DESC LIMIT 10
```

## No dictionaries encoding

The query completes in 1 second:

```shell
$ datafusion-cli -c "SELECT trace_id, MAX(time) FROM 'traces_nd.parquet' GROUP BY trace_id ORDER BY MAX(time) DESC LIMIT 10"
DataFusion CLI v30.0.0
+----------------------------------+-----------------------------+
| trace_id                         | MAX(traces_nd.parquet.time) |
+----------------------------------+-----------------------------+
| 000000000000000038e297274d174adc | 2023-07-12T18:00:07.810307  |
| 00000000000000000e577b1c8d10a199 | 2023-07-12T18:00:07.801908  |
| 0000000000000000508650f3aa158415 | 2023-07-12T18:00:04.530134  |
| 00000000000000006eb9ebf6a5d98618 | 2023-07-12T18:00:04.512612  |
| 000000000000000001c045ec504c7a90 | 2023-07-12T17:57:29.495888  |
| 00000000000000001a594b08955f5c39 | 2023-07-12T17:57:29.491154  |
| 00000000000000006f6f658ef7e81179 | 2023-07-10T06:16:34.343527  |
| 000000000000000012ad3cd4d96e6583 | 2023-07-10T06:16:34.339103  |
| 00000000000000007b6a40833cb58ae3 | 2023-07-10T06:11:34.361362  |
| 0000000000000000775bc897341290e3 | 2023-07-10T06:11:34.357441  |
+----------------------------------+-----------------------------+
10 rows in set. Query took 1.014 seconds.
```

## With dictionary encoding (query does not finish)
With dictionary encoding (the default in IOx) the query never finish (it consumes to much memory and is OOM killed):

```shell
$ datafusion-cli -c "SELECT trace_id, MAX(time) FROM 'traces.parquet' GROUP BY trace_id ORDER BY MAX(time) DESC LIMIT 10"
DataFusion CLI v30.0.0
Killed
```

When I did some basic profiling almost all the data is going to interning dictionary values in the row converter:

![Screenshot 2023-08-24 at 2 11 47 PM](https://github.com/influxdata/influxdb_iox/assets/490673/2b343095-4d15-4294-9b2a-f6e621bb2ae9)

**Comment by JayjeetAtGithub** [comment]
This is a great analysis @alamb. Thanks a lot.

**Comment by alamb** [comment]
I also wanted to see how dictionary encoding affected low cardinality columns. To do so I made a `traces_nd_service.parquet` and extracted the low cardinality (6 values)`"service.name"` tag column to a String, but leaving the other columns alone

| Query | `"service.name"` as  Dictionary | `"service.name"` as String |
|--------|--------|--------|
| `select * from ... order by "service.name" limit 10;` | 3.5s | 3.5s |
| `select max(time), "service.name" from 'traces.parquet' group by "service.name"` | 75ms | 100ms |


Thus I conclude dictionary encoding low cardinality string columns does actually help performance.

<details><summary>Details</summary>
<p>




```
(arrow_dev) alamb@MacBook-Pro-8:~/Downloads$ datafusion-cli  -c "select max(time), \"service.name\" from 'traces.parquet' group by \"service.name\"";
DataFusion CLI v29.0.0
+----------------------------+--------------+
| MAX(traces.parquet.time)   | service.name |
+----------------------------+--------------+
| 2023-07-07T04:44:30.136720 | redis        |
| 2023-07-07T04:44:30.311727 | route        |
| 2023-07-07T04:44:29.929541 | driver       |
| 2023-07-07T04:44:29.655916 | mysql        |
| 2023-07-12T18:00:07.810307 | frontend     |
| 2023-07-07T04:44:29.655817 | customer     |
+----------------------------+--------------+
6 rows in set. Query took 0.072 seconds.
```

```
(arrow_dev) alamb@MacBook-Pro-8:~/Downloads$ datafusion-cli  -c "select max(time), \"service.name\" from 'traces_nd_service.parquet' group by \"service.name\"";
DataFusion CLI v29.0.0
...
6 rows in set. Query took 0.098 seconds.
```

Here is the SQL for reproducibility
```sql

set datafusion.execution.parquet.compression='ZSTD(10)';

COPY (select attributes, duration_nano, end_time_unix_nano, "service.name"::varchar as "service.name", "span.kind", "span.name", span_id as "span_id", time, trace_id as "trace_id", "otel.status_code", parent_span_id from 'traces') to 'traces_nd_service.parquet';

❯ describe 'traces_nd_service.parquet';
+--------------------+-----------------------------+-------------+
| column_name        | data_type                   | is_nullable |
+--------------------+-----------------------------+-------------+
| attributes         | Utf8                        | YES         |
| duration_nano      | Int64                       | YES         |
| end_time_unix_nano | Int64                       | YES         |
| service.name       | Utf8                        | YES         |
| span.kind          | Utf8                        | YES         |
| span.name          | Utf8                        | YES         |
| span_id            | Dictionary(Int32, Utf8)     | YES         |
| time               | Timestamp(Nanosecond, None) | NO          |
| trace_id           | Dictionary(Int32, Utf8)     | YES         |
| otel.status_code   | Utf8                        | YES         |
| parent_span_id     | Utf8                        | YES         |
+--------------------+-----------------------------+---------

❯ select distinct "service.name" from 'traces_nd_service.parquet';
+--------------+
| service.name |
+--------------+
| redis        |
| route        |
| driver       |
| mysql        |
| customer     |
| frontend     |
+--------------+
6 rows in set. Query took 0.076 seconds.


</p>
</details> 


```

**Comment by alamb** [comment]
I verified that the difference in performance of dictionary columns is not a measurement error by making a synthetic dataset 50x larger. Dictionary encoding takes `1.68 seconds` and string encoding takes `2.5 seconds`

```shell
$ mkdir traces_nd_service
$ for i in `seq 1 50` ; do cp traces_nd_service.parquet traces_nd_service/$i.parquet ; done

$ mkdir traces_dict_service
$ for i in `seq 1 50` ; do cp traces.parquet traces_dict_service/$i.parquet ; done
```

```
(arrow_dev) alamb@MacBook-Pro-8:~/Downloads$ datafusion-cli  -c "select max(time), \"service.name\" from 'traces_nd_service' group by \"service.name\"";
DataFusion CLI v29.0.0
+-----------------------------+--------------+
| MAX(traces_nd_service.time) | service.name |
+-----------------------------+--------------+
| 2023-07-07T04:44:30.136720  | redis        |
| 2023-07-07T04:44:30.311727  | route        |
| 2023-07-07T04:44:29.655817  | customer     |
| 2023-07-07T04:44:29.929541  | driver       |
| 2023-07-07T04:44:29.655916  | mysql        |
| 2023-07-12T18:00:07.810307  | frontend     |
+-----------------------------+--------------+
6 rows in set. Query took 2.503 seconds.

(arrow_dev) alamb@MacBook-Pro-8:~/Downloads$ datafusion-cli  -c "select max(time), \"service.name\" from 'traces_dict_service' group by \"service.name\"";
DataFusion CLI v29.0.0
+-------------------------------+--------------+
| MAX(traces_dict_service.time) | service.name |
+-------------------------------+--------------+
| 2023-07-07T04:44:29.655817    | customer     |
| 2023-07-12T18:00:07.810307    | frontend     |
| 2023-07-07T04:44:30.136720    | redis        |
| 2023-07-07T04:44:30.311727    | route        |
| 2023-07-07T04:44:29.929541    | driver       |
| 2023-07-07T04:44:29.655916    | mysql        |
+-------------------------------+--------------+
6 rows in set. Query took 1.683 seconds.
```
